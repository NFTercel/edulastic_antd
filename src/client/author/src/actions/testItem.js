import { CREATE_TEST_ITEM_REQUEST, UPDATE_TEST_ITEM_REQUEST, CHECK_ANSWER, SHOW_ANSWER } from "../constants/actions";

export const createTestItemAction = data => ({
  type: CREATE_TEST_ITEM_REQUEST,
  payload: data
});

export const updateTestItemByIdAction = (id, data) => ({
  type: UPDATE_TEST_ITEM_REQUEST,
  payload: { id, data }
});

export const checkAnswerAction = mode => ({
  type: CHECK_ANSWER,
  payload: mode
});

export const showAnswerAction = mode => ({
  type: SHOW_ANSWER,
  payload: mode
});
