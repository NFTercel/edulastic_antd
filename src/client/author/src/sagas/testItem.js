import { takeEvery, call, put, all, select } from "redux-saga/effects";
import { message } from "antd";
import { isArray, isNull, isBoolean } from "lodash";
import { testItemsApi } from "@edulastic/api";
import { evaluateItem } from "../utils/evalution";
import createShowAnswerData from "../utils/showAnswer";

import {
  CREATE_TEST_ITEM_REQUEST,
  CREATE_TEST_ITEM_ERROR,
  CREATE_TEST_ITEM_SUCCESS,
  UPDATE_TEST_ITEM_REQUEST,
  UPDATE_TEST_ITEM_SUCCESS,
  UPDATE_TEST_ITEM_ERROR,
  SHOW_ANSWER,
  CHECK_ANSWER,
  ADD_ITEM_EVALUATION
} from "../constants/actions";

import { history } from "../../../configureStore";
import { getQuestionsSelector } from "../../sharedDucks/questions";

function* createTestItemSaga({ payload }) {
  try {
    const item = yield call(testItemsApi.create, payload);
    yield put({
      type: CREATE_TEST_ITEM_SUCCESS,
      payload: { item: item.data }
    });
    yield call(history.push, `/author/items/${item._id}/item-detail`);
  } catch (err) {
    console.error(err);
    const errorMessage = "Create item is failed";
    yield call(message.error, errorMessage);
    yield put({
      type: CREATE_TEST_ITEM_ERROR,
      payload: { error: errorMessage }
    });
  }
}

function* updateTestItemSaga({ payload }) {
  try {
    const item = yield call(testItemsApi.update, payload);
    yield put({
      type: UPDATE_TEST_ITEM_SUCCESS,
      payload: { item }
    });
  } catch (err) {
    console.error(err);
    const errorMessage = "Update item is failed";
    yield call(message.error, errorMessage);
    yield put({
      type: UPDATE_TEST_ITEM_ERROR,
      payload: { error: errorMessage }
    });
  }
}

function* evaluateAnswers() {
  try {
    const answers = yield select(state => state.answers);
    const validations = yield select(getQuestionsSelector);
    const { evaluation, score, maxScore } = yield evaluateItem(answers, validations);

    yield put({
      type: ADD_ITEM_EVALUATION,
      payload: {
        ...evaluation
      }
    });
    message.config({
      maxCount: 1
    });
    message.success(`score: ${score}/${maxScore}`);
  } catch (err) {
    console.error(err);
    const errorMessage = "Answer Evaluation Failed";
    yield call(message.error, errorMessage);
  }
}

function* showAnswers() {
  try {
    const answers = yield select(state => state.answers);
    const validations = yield select(getQuestionsSelector);
    const { evaluation } = createShowAnswerData(validations, answers);
    yield put({
      type: ADD_ITEM_EVALUATION,
      payload: {
        ...evaluation
      }
    });
  } catch (err) {
    console.error(err);
    const errorMessage = "Show Answer Failed";
    yield call(message.error, errorMessage);
  }
}

export default function* watcherSaga() {
  yield all([
    yield takeEvery(CREATE_TEST_ITEM_REQUEST, createTestItemSaga),
    yield takeEvery(UPDATE_TEST_ITEM_REQUEST, updateTestItemSaga),
    yield takeEvery(CHECK_ANSWER, evaluateAnswers),
    yield takeEvery(SHOW_ANSWER, showAnswers)
  ]);
}
