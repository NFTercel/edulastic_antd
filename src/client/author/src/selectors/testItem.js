import { createSelector } from "reselect";

export const stateSelector = state => state.testItem;

export const getTestItemSelector = createSelector(
  stateSelector,
  state => state.item
);
export const getTestItemCreatingSelector = createSelector(
  stateSelector,
  state => state.creating
);
export const getTestItemErrorSelector = createSelector(
  stateSelector,
  state => state.createError
);
