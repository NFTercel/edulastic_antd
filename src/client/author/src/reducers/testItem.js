import { CREATE_TEST_ITEM_REQUEST, CREATE_TEST_ITEM_SUCCESS, CREATE_TEST_ITEM_ERROR } from "../constants/actions";

const initialState = {
  item: [],
  createError: null,
  creating: false
};

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case CREATE_TEST_ITEM_REQUEST:
      return { ...state, creating: true };
    case CREATE_TEST_ITEM_SUCCESS:
      return {
        ...state,
        creating: false,
        item: payload.item
      };
    case CREATE_TEST_ITEM_ERROR:
      return { ...state, creating: false, createError: payload.error };
    default:
      return state;
  }
};
