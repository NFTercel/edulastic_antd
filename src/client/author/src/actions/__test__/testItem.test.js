import { createTestItemAction, updateTestItemByIdAction, checkAnswerAction, showAnswerAction } from "../testItem";

describe("test item", () => {
  it("create test item should return an action", () => {
    expect(
      createTestItemAction({
        rows: [
          {
            tabs: [],
            dimension: "100%",
            widgets: []
          }
        ]
      })
    ).toMatchSnapshot();

    expect(
      createTestItemAction({
        rows: [
          {
            tabs: [],
            dimension: "100%",
            widgets: [
              {
                widgetType: "question",
                type: "multipleChoice",
                title: "Multiple choice",
                reference: "5c0f0c9ed8c7f4453f5bb51f",
                tabIndex: 0
              }
            ]
          }
        ],
        columns: []
      })
    ).toMatchSnapshot();
  });

  it("update test item by id should return an action", () => {
    expect(
      updateTestItemByIdAction(1, {
        rows: [
          {
            tabs: [],
            dimension: "100%",
            widgets: []
          }
        ]
      })
    ).toMatchSnapshot();

    expect(
      updateTestItemByIdAction(2, {
        rows: [
          {
            tabs: [],
            dimension: "100%",
            widgets: [
              {
                widgetType: "question",
                type: "multipleChoice",
                title: "Multiple choice",
                reference: "5c0f0c9ed8c7f4453f5bb51f",
                tabIndex: 0
              }
            ]
          }
        ],
        columns: []
      })
    ).toMatchSnapshot();
  });

  it("check answer should return an action", () => {
    expect(checkAnswerAction()).toMatchSnapshot();
  });

  it("show answer should return an action", () => {
    expect(showAnswerAction()).toMatchSnapshot();
  });
});
