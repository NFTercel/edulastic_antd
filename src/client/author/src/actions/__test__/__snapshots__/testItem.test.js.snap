// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test item check answer should return an action 1`] = `
Object {
  "type": "[testItem] evaluate test item",
}
`;

exports[`test item create test item should return an action 1`] = `
Object {
  "payload": Object {
    "rows": Array [
      Object {
        "dimension": "100%",
        "tabs": Array [],
        "widgets": Array [],
      },
    ],
  },
  "type": "[testItem] create item request",
}
`;

exports[`test item create test item should return an action 2`] = `
Object {
  "payload": Object {
    "columns": Array [],
    "rows": Array [
      Object {
        "dimension": "100%",
        "tabs": Array [],
        "widgets": Array [
          Object {
            "reference": "5c0f0c9ed8c7f4453f5bb51f",
            "tabIndex": 0,
            "title": "Multiple choice",
            "type": "multipleChoice",
            "widgetType": "question",
          },
        ],
      },
    ],
  },
  "type": "[testItem] create item request",
}
`;

exports[`test item show answer should return an action 1`] = `
Object {
  "type": "[testItem] show test item answer",
}
`;

exports[`test item update test item by id should return an action 1`] = `
Object {
  "payload": Object {
    "data": Object {
      "rows": Array [
        Object {
          "dimension": "100%",
          "tabs": Array [],
          "widgets": Array [],
        },
      ],
    },
    "id": 1,
  },
  "type": "[testItem] update by id request",
}
`;

exports[`test item update test item by id should return an action 2`] = `
Object {
  "payload": Object {
    "data": Object {
      "columns": Array [],
      "rows": Array [
        Object {
          "dimension": "100%",
          "tabs": Array [],
          "widgets": Array [
            Object {
              "reference": "5c0f0c9ed8c7f4453f5bb51f",
              "tabIndex": 0,
              "title": "Multiple choice",
              "type": "multipleChoice",
              "widgetType": "question",
            },
          ],
        },
      ],
    },
    "id": 2,
  },
  "type": "[testItem] update by id request",
}
`;
